Match AI2-THOR's step kwargs support

mattdeitke opened the pull request:
Currently when running the following in AllenAct
```python
self.env.step(...)
```
you have to pass in a dict:
```python
self.env.step(dict(action="MoveAhead", moveMagnitude=0.25))
```
Similar to AI2-THOR, this now allows for either a dict to be passed in (i.e., it's fully backwards compatible), or just the kwargs to be passed in, as in:
```python
self.env.step(action="MoveAhead", moveMagnitude=0.25)
```

jordis-ai2 commented on allenact_plugins/ithor_plugin/ithor_environment.py:
```diff
@@ -639,7 +639,7 @@ def _snap_agent_to_initially_reachable(self, verbose=True):
 
     def step(
-        self, action_dict: Dict[str, Union[str, int, float]]
+        self, **action_dict: Dict[str, Union[str, int, float]]
```
shouldn't the new type be just `Union[str, int, float]`?

mattdeitke replied:
I don't think so. If one passes in
```python
self.env.step(action="MoveAhead", moveMagnitude=0.25)
```
`action_dict` now gets interpreted as:
```python
action_dict = dict(action="MoveAhead", moveMagnitude=0.25)
```
But one shouldn't be able to just pass in
```python
self.env.step("MoveAhead", 0.25)
```
or something like this, without the keywords.

jordis-ai2 replied:
Maybe I'm misinterpreting it, but shouldn't the type for the kwargs not be a Dict?
[pep-0484](https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values)

mattdeitke replied:
Sorry, missed this. I think you're right, `**kwargs` should be annotated with the types and not as a dict. Fixed in https://github.com/allenai/allenact/pull/290/commits/33a8f40da7aa924c3e3682c1edd3b002f421ec02.

jordis-ai2 commented on allenact_plugins/robothor_plugin/robothor_environment.py:
```diff
@@ -460,7 +460,7 @@ def last_action_return(self) -> Any:
         return self.controller.last_event.metadata["actionReturn"]
 
-    def step(self, action_dict: Dict) -> ai2thor.server.Event:
+    def step(self, **action_dict: Dict) -> ai2thor.server.Event:
```
(same as above)

jordis-ai2 reviewed (commented):
LGTM. I'd just update the types for consistency 👍

Lucaweihs commented:
As the LGTM.com checks call out: this change would break lots of things. You should instead do something something like
``python
def step(self, action_dict: Optional[Dict[str, Union[str, int, float]]] = None, **kwargs: Dict[str, Union[str, int, float]]):
    if action_dict is None:
        action_dict = {}
    action_dict.update(kwargs)
    ...
```

mattdeitke commented:
Hmm.. Interesting.. Ok. I thought I recalled that passing in a dict to `**kwargs` would work, but perhaps that's only if one uses `kwargs=dict(...)`.

mattdeitke commented:
Updated and tested in: https://colab.research.google.com/drive/1bKe9uGeJuXkVPhR3kMIJlWme9KnLjG5d?usp=sharing

Note that
```python
def step(
    self,
    action_dict: Optional[Dict[str, Union[str, int, float]]] = None,
    **kwargs: Dict[str, Union[str, int, float]]
):
    if action_dict is None:
        action_dict = {}
    action_dict.update(kwargs)
```
was simplified to
```python
def step(
    self,
    action_dict: Optional[Dict[str, Union[str, int, float]]] = dict(),
    **kwargs: Dict[str, Union[str, int, float]]
):
    action_dict.update(kwargs)
```

Lucaweihs commented:
I know you're reverted things but, as this is a common python gotcha, I should stress that doing
```python
    action_dict: Optional[Dict[str, Union[str, int, float]]] = dict(),
```
**can lead to absolutely disastrous consequences** and is not at all equivalent to doing
```python
    action_dict: Optional[Dict[str, Union[str, int, float]]] = None,
```

To see why this is the case, try running the following:
```
def test(yar = dict()):
    yar["a"] = yar.get("a", 0) + 1
    print(yar)

for i in range(10):
    test()
```
What you'll see is that the `yar` dictionary is persistent across calls to `test`!

mattdeitke commented:
Oh, wow. Interesting. I imagined `yar = dict()` created a new dict instance each time `test()` was called.

Lucaweihs commented:
@mattdeitke I'm happy with this being merged once you've addressed Jordi's type note above.

mattdeitke commented:
Fixed. Also for consistency, in https://github.com/allenai/allenact/pull/290/commits/272d3e86ea85e1d98e1094106a5a52c1860d003f I've added a dict as a possible value for `action_dict`, since some actions like
```python
controller.step(action="Teleport", rotation=dict(x=0, y=0, z=0))
```
